src/components/account/account.ts

```diff
@@ -72,6 +48,135 @@ export async function postUseGoogleSSO(ctx: IContext, _params: IParameters): Pro
       linkTo: ctx.linkTo,
       csrf: ctx.csrf,
       location: ctx.app.location,
+      providerName: 'Google',
+    }),
+  };
+}
+
+export async function getUseMicrosoftSSO(ctx: IContext, _params: IParameters): Promise<IResponse> {
+  const cfgProvided = ctx.app.oidcProviders.find((p: IOIDCConfig) => {
+    return p.providerName === 'microsoft';
+  });
+
+  if (!cfgProvided) {
+    ctx.app.logger.error('Unable to find Microsoft OIDC config');
+
+    return {
+      body: error500.render({
+        routePartOf: ctx.routePartOf,
+        linkTo: ctx.linkTo,
+        csrf: ctx.csrf,
+        location: ctx.app.location,
+      }),
+    };
+  }
+
+  const user = await fetchLoggedInUser(ctx);
+
+  return {
+    body: useMicrosoftSSOTemplate.render({
+      routePartOf: ctx.routePartOf,
+      linkTo: ctx.linkTo,
+      csrf: ctx.csrf,
+      location: ctx.app.location,
+      user,
+    }),
+  };
+}
+
+export async function postUseMicrosoftSSO(ctx: IContext, _params: IParameters): Promise<IResponse> {
+  const cfgProvided = ctx.app.oidcProviders.find((p: IOIDCConfig) => {
+    return p.providerName === 'microsoft';
+  });
+
+  if (!cfgProvided) {
+    ctx.app.logger.error('Unable to find Microsoft OIDC config');
+
+    return {
+      body: error500.render({
+        routePartOf: ctx.routePartOf,
+        linkTo: ctx.linkTo,
+        csrf: ctx.csrf,
+        location: ctx.app.location,
+      }),
+    };
+  }
+
+  const oidcClient = new OIDC(
+    cfgProvided.clientID,
+    cfgProvided.clientSecret,
+    cfgProvided.discoveryURL,
+    ctx.linkTo('account.use-microsoft-sso-callback.get', undefined, true),
+  );
+  const redirectURL = await oidcClient.getAuthorizationOIDCURL(ctx.session);
+
+  return {
+    redirect: redirectURL,
+  };
+}
+
+export async function getMicrosoftOIDCCallback(ctx: IContext, _params: IParameters): Promise<IResponse> {
+  const uaa = new UAAClient({
+    apiEndpoint: ctx.app.uaaAPI,
+    clientCredentials: {
+      clientID: ctx.app.oauthClientID,
+      clientSecret: ctx.app.oauthClientSecret,
+    },
+  });
+
+  const cfgProvided = ctx.app.oidcProviders.find((p: IOIDCConfig) => {
+    return p.providerName === 'microsoft';
+  });
```

Comment:
Again, could be a bit shorter. I wonder whether it would be worth making `oidcProviders` a `Map<'google'|'microsoft', IOIDCConfig>` instead of an array that we have to keep searching? 🤔

Reply:
Maybe `Map<string, IOIDCConfig>` if you don't want the type system to pretend it's not possible to configure other providers, although as it stands there would be no point in configuring anything other than microsoft because the code always goes fishing by providerName.